src/string/extglobmatch.c

```
@@ -0,0 +1,276 @@
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "extglobmatch.h"
+#include "brz.h"
+
+#define cat_one()    do{                                    \
+                        if (atoms > 1) {                    \
+                            --atoms;                        \
+                            glob *b = *--frag;              \
+                            glob *a = *--frag;              \
+                            *frag++ = glob_seq(a,b);             \
+                            free_glob(a);                   \
+                            free_glob(b);                   \
+                        }                                   \
+                    } while(0)
+
+#define cat_all()    do{                                    \
+                          while (--atoms > 0) {             \
+                              glob *b = *--frag;            \
+                              glob *a = *--frag;            \
+                              *frag++ = glob_seq(a,b);           \
+                              free_glob(a);                 \
+                              free_glob(b);                 \
+                          }                                 \
+                    }while(0)
+
+static glob* match_glob = NULL;
+
+static void cleanup(glob **start, glob **end) {
+    while (start != end)
+    {
+        free_glob(*start);
+        ++start;
+    }
+}
+
+static void print_error(char const c) {
+    fprintf(stderr, "pattern error: Unexpected character '%c'\n", c);
+}
+
+static glob* handle_operator(char op, glob const* g)
+{
+    glob* tmp = NULL;
+    switch (op) {
+        case '!': tmp = glob_not(g);
+                  break;
+        case '*': tmp = glob_star(g);
+                  break;
+        case '+': tmp = glob_plus(g);
+                  break;
+        case '@': tmp = glob_at(g);
+                  break;
+        case '?': tmp = glob_opt(g);
+                  break;
+    }
+    return tmp;
```

Reaching this should yield a hard abort(), because there is no way that this can be caused by the user. No need to propagate this error any further.